ServerCore/Pages/Puzzles/PlayPuzzles.cshtml.cs

```diff
@@ -0,0 +1,237 @@
+            var singlePlayerPuzzleStatePerPlayer = _context.SinglePlayerPuzzleStatePerPlayer
+                .Where(state => state.PlayerID == LoggedInUser.ID && state.Puzzle.EventID == Event.ID);
+            foreach (SinglePlayerPuzzleStatePerPlayer statePerPlayer in singlePlayerPuzzleStatePerPlayer)
+            {
+                singlePlayerPuzzleViewDict[statePerPlayer.PuzzleID].SolvedTime = statePerPlayer.SolvedTime;
+                singlePlayerPuzzleViewDict[statePerPlayer.PuzzleID].UnlockedTime = statePerPlayer.UnlockedTime;
+            }
+
+            IEnumerable<PuzzleView> visibleSinglePlayerPuzzlesQ = singlePlayerPuzzleViewDict.Values.AsEnumerable().Where(puzzleView => ShowAnswers || puzzleView.UnlockedTime != null);
+            visibleSinglePlayerPuzzlesQ = this.GetSortedView(visibleSinglePlayerPuzzlesQ, sortOrder ?? DefaultSort);
+            if (StateFilter == PuzzleStateFilter.Unsolved)
+            {
+                visibleSinglePlayerPuzzlesQ = visibleSinglePlayerPuzzlesQ.Where(puzzles => puzzles.SolvedTime == null);
+            }
+
+            return visibleSinglePlayerPuzzlesQ;
+        }
+
+        private IEnumerable<PuzzleView> GetSortedView(IEnumerable<PuzzleView> puzzleViews, SortOrder? sortOrder)
+        {
+            if (sortOrder == null)
+            {
+                return puzzleViews;
+            }
+
+            switch (sortOrder)
+            {
+                case SortOrder.PuzzleAscending:
+                    return puzzleViews.OrderBy(pv => pv.Name);
+                case SortOrder.PuzzleDescending:
+                    return puzzleViews.OrderByDescending(pv => pv.Name);
+                case SortOrder.GroupAscending:
+                    return puzzleViews.OrderBy(pv => pv.Group).ThenBy(pv => pv.OrderInGroup).ThenBy(pv => pv.Name);
+                case SortOrder.GroupDescending:
+                    return puzzleViews.OrderByDescending(pv => pv.Group).ThenByDescending(pv => pv.OrderInGroup).ThenByDescending(pv => pv.Name);
+                case SortOrder.SolveAscending:
+                    return puzzleViews.OrderBy(pv => pv.SolvedTime ?? DateTime.MaxValue);
+                case SortOrder.SolveDescending:
+                    return puzzleViews.OrderByDescending(pv => pv.SolvedTime ?? DateTime.MaxValue);
+                default:
+                    throw new ArgumentException($"unknown sort: {sortOrder}");
+            }
+        }
+
+        private async Task PopulatePuzzleViewsWithFiles(
+            IEnumerable<PuzzleView> puzzleViews,
+            Dictionary<int, ContentFile> files)
+        {
+            foreach (var puzzleView in puzzleViews)
+            {
+                files.TryGetValue(puzzleView.ID, out ContentFile content);
+                puzzleView.Content = content;
+            }
+
+            if (ShowAnswers)
+            {
+                Dictionary<int, ContentFile> answers = await(from file in _context.ContentFiles
```

Comment:
No reason to do this twice - precompute this in OnGetAsync (it can be null if ShowAnswers is false) and pass as it an argument.

Reply:
Fixed in newest iteration